rocAL/source/readers/image/numpy_data_reader.cpp

```diff
@@ -0,0 +1,642 @@
+/*
+Copyright (c) 2025 Advanced Micro Devices, Inc. All rights reserved.
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+*/
+
+#include <algorithm>
+#include <cassert>
+#include <cstring>
+#include <numeric>
+#include <random>
+#include <math.h>
+#include "pipeline/commons.h"
+#include "pipeline/filesystem.h"
+#include "readers/image/numpy_data_reader.h"
+
+// the HEADER_OFFSET is a magic number - the first 10 bytes store info about numpy version and header len
+// The first 6 bytes are a magic string: exactly \x93NUMPY.
```

Comment:
why do you need a do,while loop here. Looks like the func is only called once. I think the following would code will be more efficient
I think you can get rid of using this function as it is just calling func. replace it with "func;" and check for "_header_parsing_failed" flag after